falcon/util/uri.py

```diff
@@ -292,9 +292,9 @@ def parse_query_string(query_string, keep_blank_qs_values=False):
             # a list, or append the next value to the list.
             old_value = params[k]
             if isinstance(old_value, list):
-                old_value.append(v)
+                old_value.append(decode(v))
             else:
-                params[k] = [old_value, v]
+                params[k] = [old_value, decode(v)]
 
         else:
             if ',' in v:
```

Comment:
>> Seems like v must be decoded before checking if it contains commas, since the
>> commas might be encoded. What do you think about factoring out the decode(v)
>> and placing it above L290, e.g.:

This is incorrect.  The purpose of the % encoding is to "turn off the special nature" of the character.  It's just like backslashing a character in a C string.  So if we decode prior to splitting, we are in trouble.

The procedure should be:
1. Split on literal "&" to split the n=v parameter pairs.
2. For each of these parameter pairs, subsplit on ",", which is simply the next delimiter.
3. Now decode each fragment separately.  This means:
    a) decode the name of a parameter pair (e.g. a utf-8 funny char in an identifier...god help us)
    b) decode the value:
        i) Single value
       ii) Multiple values (split on ","), each of which must be decoded.

3.b.ii. also shows how someone needs to do encoding if they have multi-valued parameters separated by "," (they must encode each part separately, and NOT encode the ",").  Note that we are EXTENDING the meaning of "," to have a special meaning in the same spirit that the original spec had of "&".  "," special meaning is - turn this value into an array, and use "," as the subdelimiter.
   

v = decode(v)

if k in params:
    # ...